### DISCUSSION TITLE
Make the install a bit more portable

### DISCUSSION THREAD

**Discussion by fizion** [rsxdalv/TTS-WebUI#76] [Ideas]
Like said before, I'm a noob and not that handy with all commands to install certain things using pip etc. 

My Windows had some weird issues and since it automatically creates an image of C, I decided to put an image back from a week before. 

That was before I actually installed the TTS Generation WebUI which is basically located on drive G but also has a lot of files on C as I noticed. I tried to get them back from yesterdays image but there must be some directories/files somewhere it misses since it no longer works. I get a lot of things like below when I press Generate in in the Bark tab.

```
Generating: ''' hello '''
Parameters: history_prompt: None text_temp: 0.7 waveform_temp: 0.7
Traceback (most recent call last):
  File "G:\tts\installer_files\env\lib\site-packages\gradio\routes.py", line 437, in run_predict
    output = await app.get_blocks().process_api(
  File "G:\tts\installer_files\env\lib\site-packages\gradio\blocks.py", line 1352, in process_api
    result = await self.call_function(
```

Totally my problem but I think it would be nice if the script actually would ask where you would want to put all extra files. So basically at the end you've one directory which has all so it's also clear which files belong to what and I don't ruin a Stable Diffusion thing as well ;)

A lot of files are stored in %appdata%/Python/Python310 I think and all models are in the .cache directory under the user. Especially the .cache directory uses a lot of space which I would like to have on G, together with the other files. That will make loading models much slower for sure but my NVme has a limit capacity on C.

Having all things at one location would make it clear what you can delete if you want to start over (like me) but also it will be easier to move things around. F.e. I also moved my Stable Diffusion to G but that had a few lines in a config file which you could change so you could move the models etc. along with the rest.

Maybe an idea to have some things implemented so it's easier to see what will end up where and change it afterwards or, maybe better, on install.

I tried already installing things again but think it still uses old files somewhere it shouldn't use or it misses some. The Install buttons mentioned at my earlier discussion post don't seem to work in last version and I saw the install button for Demucs Demo was called 'MusicGen', not a biggie ofcourse.

For now all exclamation marks in the tabs are gone and only Bark doesn't seem to work. And, man, how bizar is that Demucs, that's impressive!

**Comment by rsxdalv** [comment]
Related to https://github.com/rsxdalv/tts-generation-webui/issues/68

**Reply by fizion** [reply]
Sorry totally missed that one. And I indeed saw there was a way to move for example the hugging-face cache to a different location. I was actually thinking if it would do harm to just copy .cache to a directory of your likings and in Windows just create a symbolic link pointing .cache to that directory. Not the cleanest way I think. 

Stupid comparison but a lot of people do it with their FlightSimulator addons since those are so huge and if you have them all active loading takes forever. So someone created a little application which just creates symbolic links in the actual directory where addons should be so you don't have to move over massive amount of files every time you fly in a different area.

**Reply by rsxdalv** [reply]
A symlink sounds interesting, that might be a workaround. I wouldn't put it into this project since that would **absolutely** affect everything on your system; who knows how soon we will have desktop apps that rely on this .cache., but as a power-user tool, it could be a good way to deal with the hundreds of gigabytes of model weights.

**Reply by fizion** [reply]
I've played another afternoon and moved the cache folder to another drive. I made a junction instead of symlink but I don't think it matters much in this case. 

For those who like to try, just move your .cache content to another location (in my example G:\[cache]). The moving of items can hang on the symbolic links in some huggingface directories so I would remove them first. If it's all moved and your .cache directory is empty, remove it. Next you put something like this in your command box:

`mklink /J "C:\Users\[username]\.cache" "G:\[Cache]"`

@rsxdalv for the project it's not a good idea since it will affect other things as well. For example, I see that Chaos (from the Chaos Group) also has something in there from their benchmark tool. But for those who just want to move their huge models it might be an idea to just do it for the suno and huggingface directory. Here those are the actual large directories.

**Reply by rsxdalv** [reply]
Great work! I added some of the environment variables to manage these within the UI; however there will be other projects that could cause diplicate downloads. Therefore it might be a good idea to symlink for those who know how to do it.

**Reply by fizion** [reply]
I happen to update the UI today and I immediately noticed the different settings part with submenu and the path options, nice! For now the junction here works nice. But I actually use it mainly for the .cache directory. The AI image stuff shares much more GB's worth of models and most of those UIs have the option to look in different folders as well. Getting friendlier every week ;)

**Comment by rsxdalv** [comment]
To be fair, if the model was on an unfragmented HDD, it's quite possible that it wouldn't always be the bottleneck, HDDs can get like 300mb/s on large files. And for now there hasn't been any hot-swapping with hdd, only with RAM, so shouldn't be too worrysome.

**Reply by fizion** [reply]
True, but since I moved my Stable Diffusion it did get a lot slower. Maybe especially the initial loading. I guess because there are a lot small files involved which NVMe's are good at.

**Reply by rsxdalv** [reply]
Yeah totally, if you are talking about the installation itself, it would be faster if it's on NVMe, now the model weights are not as important, also since you normally cache them.

**Comment by rsxdalv** [comment]
The thing about "managing" these files is that a) there are about 5+ different methods files (weights) are acquired, and it's not always _proper_ from license standpoint to say "copy" everything to HuggingFace. For example, until recently, some of the model downloads were hosted on Facebook's own servers.
b) these models are shared between projects. if you delete them from somewhere you might have to re-download it for another webui.

What's important to note is also that if you just wanted a "clean install" then generally you would only need to install the core code/dependencies without all of the model weights.

Basically that's why the default _settings_ (if you can call them that) are a bit flaky. They have benefits and drawbacks, but this way, it's less likely that you have gigabytes of unneeded files.

Finally, to answer the question - how to "start over" - assuming everyone's implementations are sane (SD/text generation etc etc) you can just delete and redownload the caches. It should not break the stuff.

**Reply by fizion** [reply]
That makes sense indeed. I'm also sure there is a logic behind it all, it's just I don't get it always. Being from the early 70's I've had my share of DOS but the entire github logic is still a bit mystic to me.
